File: tp1/python/src/etudiant.py
from src.personne import Personne
from src.cours import Cours

class Etudiant(Personne):
    def __init__(self, nom, age, numero, moyenne):
        super().__init__(nom, age)
        self.__numero = numero
        self.moyenne = moyenne
        self.__cours = []

    @property
    def numero(self):
        return self.__numero

    @property
    def moyenne(self):
        return self.__moyenne

    @moyenne.setter
    def moyenne(self, value):
        if value < 0 or value > 20:
            raise ValueError("La moyenne doit être comprise entre 0 et 20")
        self.__moyenne = value

    def ajouter_cours(self, cours):
        self.__cours.append(cours)

    def afficher_details(self):
        super().afficher_details()
        print(f"Numero etudiant : {self.__numero}, Moyenne : {self.__moyenne}/20")
        if self.__cours:
            print("Cours inscrits :")
            for c in self.__cours:
                c.afficher_details()

File: tp1/java/main/java/heritage/Etudiant.java
package heritage;

import java.util.ArrayList;
import java.util.List;

public class Etudiant extends Personne {
    private String numero;
    private double moyenne;
    private List<Cours> cours;

    public Etudiant(String nom, int age, String numero, double moyenne) {
        super(nom, age);
        this.numero = numero;
        setMoyenne(moyenne);
        this.cours = new ArrayList<>();
    }

    public String getNumero() {
        return numero;
    }

    public double getMoyenne() {
        return moyenne;
    }

    public void setMoyenne(double moyenne) {
        if (moyenne < 0 || moyenne > 20) {
            throw new IllegalArgumentException("La moyenne doit être comprise entre 0 et 20");
        }
        this.moyenne = moyenne;
    }

    public void ajouterCours(Cours cours) {
        this.cours.add(cours);
    }

    public List<Cours> getCours() {
        return cours;
    }

    @Override
    public void afficherDetails() {
        super.afficherDetails();
        System.out.println("Numero etudiant : " + numero + ", Moyenne : " + moyenne + "/20");
        if (!cours.isEmpty()) {
            System.out.println("Cours inscrits :");
            for (Cours c : cours) {
                c.afficherDetails();
            }
        }
    }
}
